src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1063 @@
+                if (result != protocol) throw new Exception($"Protocol '{protocol}' not supported by '{peer}'.");
+
+                return stream;
+            }
+            catch (Exception)
+            {
+                stream?.DisposeAsync();
+                throw;
+            }
+        }
+
+        /// <summary>
+        ///   Establish a duplex stream between the local and remote peer.
+        /// </summary>
+        /// <param name="remote"></param>
+        /// <param name="addrs"></param>
+        /// <param name="cancel"></param>
+        /// <returns></returns>
+        private async Task<PeerConnection> DialAsync(Peer remote,
+            IEnumerable<MultiAddress> addrs,
+            CancellationToken cancel)
+        {
+            log.Debug($"Dialing {remote}");
+
+            if (remote == LocalPeer) throw new Exception("Cannot dial self.");
+
+            // If no addresses, then ask peer routing.
+            if (Router != null && addrs.Count() == 0)
+            {
+                var found = await Router.FindPeerAsync(remote.Id, cancel).ConfigureAwait(false);
+                addrs = found.Addresses;
+                remote.Addresses = addrs;
+            }
+
+            // Get the addresses we can use to dial the remote.  Filter
+            // out any addresses (ip and port) we are listening on.
+            var blackList = listeners.Keys
+               .Select(a => a.WithoutPeerId())
+               .ToArray();
+            var possibleAddresses =
+                (await Task.WhenAll(addrs.Select(a => a.ResolveAsync(cancel))).ConfigureAwait(false))
+               .SelectMany(a => a)
+               .Where(a => !blackList.Contains(a.WithoutPeerId()))
+               .Select(a => a.WithPeerId(remote.Id))
+               .Distinct()
+               .ToArray();
+            if (possibleAddresses.Length == 0) throw new Exception($"{remote} has no known or reachable address.");
+
+            // Try the various addresses in parallel.  The first one to complete wins.
+            PeerConnection connection = null;
+            try
+            {
+                using (var timeout = new CancellationTokenSource(TransportConnectionTimeout))
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, cancel))
+                {
+                    var attempts = possibleAddresses
+                       .Select(a => DialAsync(remote, a, cts.Token));
+                    connection = await TaskHelper.WhenAnyResultAsync(attempts, cts.Token).ConfigureAwait(false);
+                    cts.Cancel(); // stop other dialing tasks.
+                }
+            }
+            catch (Exception e)
+            {
+                var attemped = string.Join(", ", possibleAddresses.Select(a => a.ToString()));
+                log.Trace($"Cannot dial {attemped}");
+                throw new Exception($"Cannot dial {remote}.", e);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)